`generate.py` tests

diivm opened the pull request:
- Part of splitting #8 (2/N)

diivm committed:
[test_generate.py] init file

diivm committed:
[test_generate.py] intial setup and some tests

diivm committed:
[test_generate.py] fix intial_lines problem, new function

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+import test_parse
+
+
+def get_character_only_string(word):
```
Is it a word, or a line? What's the aim here? Removing space will throw readability out of whack

diivm replied:
For stringent comparison of the output and desired output.
Updated.

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+    return "".join(chars)
+
+
+def get_binded_code(tmp_path, cpp_code_block):
```
`generate_bindings`

Also swap the order since the code is more important than the location of the tmp file

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+        3. Convert the list to a string and then return the stripped off string.
+
+    Parameters:
+        - tmp_path (Path): The tmp_path for test folder
```
Use Python's type notation please

diivm replied:
Updated to `pathlib.PosixPath` :+1:

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+    utils.dump_json(filepath=json_path, info=parsed_info)
+
+    # Get the binded code
+    binded_code = generate.generate(source=json_path)
```
Is it possible for the generate to not need to read from a file? This will be an IO bottleneck otherwise

diivm replied:
Yeah, I'll be updating the `generate.py` as well this file (remember I mentioned to make JSON usage optional). This will be better done when working on #37

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+
+    # JSON dump the parsed info
+    json_path = tmp_path / "file.json"
+    utils.dump_json(filepath=json_path, info=parsed_info)
```
You may want to minimize IO by optionally minifing the json. To do that, use `seperators` in the json dump command

diivm replied:
Updated

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+        - chars (str): The stripped off string
+    """
+
+    chars = []
```
```py
chars = str([x for x in words if not x.isspace])
```

kunaltyagi replied:
`str` might not work. You may have to use `"".join([x for ...])`

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+    expected_module_code = """
+    PYBIND11_MODULE(pcl, m){
+        py::class_<AStruct>(m, "AStruct")
+        .def_readwrite("aMember", &AStruct::aMember);
```
No default ctor?

diivm replied:
Added.

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+
+def get_expected_string(expected_module_code):
+    """
+    Returns expected output string after combining inclusions and pybind11 linelist.
```
What's a linelist?

diivm replied:
Changed to `initial_pybind_lines`.

These are the lines:
```py
    initial_pybind_lines = [
        "#include <pybind11/pybind11.h>",
        "#include<pybind11/stl.h>",
        "#include<pybind11/stl_bind.h>",
        "namespace py = pybind11;",
        "using namespace py::literals;",
    ]
```

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+        - expected_output (str): The stripped off, combined code.
+    """
+
+    file_inclusion = "#include <pcl/file.cpp>"
```
`file_include`? `file_include_str`?

diivm replied:
Changed to `file_include`.

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+        - expected_output (str): The stripped off, combined code.
+    """
+
+    file_inclusion = "#include <pcl/file.cpp>"
```
Why is the `pcl` prefix needed? What about just `'#include "file.cpp"'`?

diivm replied:
Done.

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+    file_inclusion = "#include <pcl/file.cpp>"
+
+    # Get pybind11 linelist in the form of a string
+    pybind_linelist = generate.bind.pybind_linelist
```
`pybind11_preamble = "".join(generate.bind.pybind_linelist)`

kunaltyagi commented on bindings/python/tests/test_generate.py:
```diff
@@ -0,0 +1,111 @@
+import test_parse
+
+
+def get_character_only_string(word):
```
`remove_whitespace`?

diivm committed:
[test_generate.py] updates

- rename `get_character_only_string` -> `remove_whitespace`
- functional changes to `remove_whitespace`
- rename `get_binded_code` -> `generate_bindings`
- tmp_path change order and comment updation
- dump_json: separators for json size reduction
- include: remove pcl
- pybind_linelist -> initial_pybind_lines
- default constructor

diivm committed:
[utils.py] add options to dump_json

diivm committed:
[generate.py] renaming and change `pcl/filename`->`filename` in include

kunaltyagi commented on bindings/python/scripts/generate.py:
```diff
@@ -199,6 +199,9 @@ def handle_struct_decl(self):
             self.linelist.append(f'py::class_<{class_name}>(m, "{class_name}")')
 
+        # default constructor
+        self.linelist.append(".def(py::init<>())")
```
Would this cause issues if there's already a ctor before or if default ctor is deleted?

diivm replied:
Handled via an if condition: don't add default ctor in `handle_constructor`
```diff
-        self.linelist.append(f".def(py::init<{argument_type_list}>())")
+        # default ctor `.def(py::init<>())` already inserted while handling struct/class decl
+        if argument_type_list:
+            self.linelist.append(f".def(py::init<{argument_type_list}>())")
```

kunaltyagi reviewed (commented):
Tests in test_generate.py look awesome

diivm committed:
[test_generate.py] str won't work

diivm committed:
[generate.py] if condition for default ctor in handle_constructor

aPonza reviewed (approved):
Well done, the tests are so simple and clear to read. LGTM

divmadan commented:
Tests are passing for the current `generate.py`. For merging, I'll be rebasing with on the `devel/python_bindings` branch (which doesn't have `generate.py` merged yet) so CI will fail.
Rebasing and merging..
